### ISSUE TITLE
Inconsistent results

### ISSUE THREAD

**Issue by ryan-lowe** [alexis-jacq/LOLA_DiCE#1]
First of all, thank you for providing such a readable public implementation of an interesting paper!

Unfortunately, I've found (using PyTorch 4.1) that the results for `IPD_DiCE.py` seem to vary significantly between runs (you define a seed in the Hp() class, but you don't seem to set it anywhere). Out of 3 runs, 2 of them converged to (-1.9, -1.9) reward after 200 updates (the 'all defect' policy), and only one of the runs got down to about (-1.03, -1.28) reward. 

Any idea why this could be the case? Are you also experiencing this?

(LOLA with exact gradients in `IPD_ex.py` does seem to work though!)

**Comment by alexis-jacq** [comment]
Thanks for your feedback!

> you define a seed in the Hp() class, but you don't seem to set it anywhere
I reinitialize the seed before each session (with != nb_lookahead) in the __main__ script.

Besides, I have also experienced the 1-lookahead (is it what you tried?) going back to a defect strategy after ~ 300 updates. 
Anyway, I'm going to push the last version, that works properly.

**Comment by alexis-jacq** [comment]
I still have to update the picture, but you can already test the new version. I've just tested the 1-lookahead over 400 updates and different seeds, it seems to cooperate in all cases. 

Now, it's the 3-lookaheads model that seems quite poor, but it's just taking more time before starting to cooperate. 3 lookahead is probably too much with these hyperparameters (and may require smaller learning rates).

**Comment by jleni** [comment]
@alexis-jacq Thanks alexis for this implementation.
I have just run the new version and I also see some odd results for lookahead=3

![image](https://user-images.githubusercontent.com/13698520/44172785-7de2ce80-a0de-11e8-847a-e467a571b111.png)

Here is a much longer run for lookahead=3

![image](https://user-images.githubusercontent.com/13698520/44173940-bdf78080-a0e1-11e8-95bc-e7a74eba043b.png)

**Comment by jleni** [comment]
I see that you are applying a baseline. In the original code, the baseline seems be to disabled.

**Comment by alexis-jacq** [comment]
Oh, thanks a lot @jleni for the longer run! I'm glad to know it finally works for 3 lookaheads after a longer run. Yes, the baseline is disabled by default on the original code, but I think the figures on the DiCE paper were obtained with the baseline (it really improve the gradient estimation).

**Comment by jleni** [comment]
I am a bit confused with these lines:

https://github.com/alexis-jacq/LOLA_DiCE/blob/27254c684e15c11587adf42c4b17bc435ca69183/ipd_DiCE.py#L56

https://github.com/alexis-jacq/LOLA_DiCE/blob/27254c684e15c11587adf42c4b17bc435ca69183/ipd_DiCE.py#L59

Why do you add the logprobs instead of concatenating the two vectors?
It think that leads to incorrect results.

**Comment by jleni** [comment]
Ignore my comments. You are calculating tau and it is actually the same.
